components/application-authenticators/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAAuthenticator.java

```diff
@@ -70,25 +70,25 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
     @Override
     protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
                                                  AuthenticationContext context) throws AuthenticationFailedException {
-        //Get the authenticated user principle
-        Principal principal = request.getUserPrincipal();
-        if (principal == null) {
-            HttpSession session = request.getSession(false);
-            if (session != null) {
-                principal = (Principal) session
-                        .getAttribute(IWAServelet.PRINCIPAL_SESSION_KEY);
+
+        HttpSession session = request.getSession(false);
+        if (session.getAttribute(IWAConstants.SUBJECT_ATTRIBUTE)  == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Attribute not set. Therefore authentication is failed.");
             }
+            throw new AuthenticationFailedException("Authentication Failed");
         }
-
-        if (principal == null || principal.getName() == null) {
+        String username = session.getAttribute(IWAConstants.SUBJECT_ATTRIBUTE).toString();
+        if (username == null) {
```

Comment:
the toString() method would have returned NPE if the value is null. so null check should be done before calling toString()